File: 099. Recover Binary Search Tree/Python/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def recoverTree(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        pre = None
        first = None
        second = None
        while root:
            if root.left:
                temp = root.left
                while temp.right and temp.right != root:
                    temp = temp.right
                if not temp.right:
                    temp.right = root
                    root = root.left
                else:
                    temp.right = None
                    if pre and pre.val > root.val:
                        if first:
                            second = root
                        else:
                            first = pre
                            second = root
                    pre = root
                    root = root.right
            else:
                if pre and pre.val > root.val:
                    if first:
                        second = root
                    else:
                        first = pre
                        second = root
                pre = root
                root = root.right
        if first and second:
            first.val, second.val = second.val, first.val

File: 099. Recover Binary Search Tree/Java/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void recoverTree(TreeNode root) {
        TreeNode pre = null;
        TreeNode first = null;
        TreeNode second = null;
        while (root != null) {
            if (root.left != null) {
                TreeNode temp = root.left;
                while (temp.right != null && temp.right != root) {
                    temp = temp.right;
                }
                if (temp.right == null) {
                    temp.right = root;
                    root = root.left;
                } else {
                    temp.right = null;
                    if (pre != null && pre.val > root.val) {
                        if (first != null) {
                            second = root;
                        } else {
                            first = pre;
                            second = root;
                        }
                    }
                    pre = root;
                    root = root.right;
                }
            } else {
                if (pre != null && pre.val > root.val) {
                    if (first != null) {
                        second = root;
                    } else {
                        first = pre;
                        second = root;
                    }
                }
                pre = root;
                root = root.right;
            }
        }
        if (first != null && second != null) {
            int temp = first.val;
            first.val = second.val;
            second.val = temp;
        }
    }
}